core/data/clients/memory-database_test.go

```diff
@@ -0,0 +1,612 @@
+	if err != nil {
+		t.Fatalf("Error getting ValuesByValueDescriptorNames: %v", err)
+	}
+	if len(values) != 2 {
+		t.Fatalf("There should be 2 Values, not %d", len(values))
+	}
+	values, err = db.ValueDescriptorsByName([]string{"name1", "name"})
+	if err != nil {
+		t.Fatalf("Error getting ValuesByValueDescriptorNames: %v", err)
+	}
+	if len(values) != 1 {
+		t.Fatalf("There should be 1 Values, not %d", len(values))
+	}
+	values, err = db.ValueDescriptorsByName([]string{"name", "INVALID"})
+	if err != nil {
+		t.Fatalf("Error getting ValuesByValueDescriptorNames: %v", err)
+	}
+	if len(values) != 0 {
+		t.Fatalf("There should be 0 Values, not %d", len(values))
+	}
+
+	values, err = db.ValueDescriptorsByUomLabel("name1")
+	if err != nil {
+		t.Fatalf("Error getting ValuesByValueDescriptorNames: %v", err)
+	}
+	if len(values) != 1 {
+		t.Fatalf("There should be 1 Values, not %d", len(values))
+	}
+	values, err = db.ValueDescriptorsByUomLabel("INVALID")
+	if err != nil {
+		t.Fatalf("Error getting ValuesByValueDescriptorNames: %v", err)
+	}
+	if len(values) != 0 {
+		t.Fatalf("There should be 0 Values, not %d", len(values))
+	}
+
+	values, err = db.ValueDescriptorsByLabel("name1")
+	if err != nil {
+		t.Fatalf("Error getting ValuesByValueDescriptorNames: %v", err)
+	}
+	if len(values) != 1 {
+		t.Fatalf("There should be 1 Values, not %d", len(values))
+	}
+	values, err = db.ValueDescriptorsByLabel("INVALID")
+	if err != nil {
+		t.Fatalf("Error getting ValuesByValueDescriptorNames: %v", err)
+	}
+	if len(values) != 0 {
+		t.Fatalf("There should be 0 Values, not %d", len(values))
+	}
+
+	values, err = db.ValueDescriptorsByType("name1")
+	if err != nil {
+		t.Fatalf("Error getting ValuesByValueDescriptorNames: %v", err)
+	}
+	if len(values) != 1 {
+		t.Fatalf("There should be 1 Values, not %d", len(values))
+	}
+	values, err = db.ValueDescriptorsByType("INVALID")
+	if err != nil {
+		t.Fatalf("Error getting ValuesByValueDescriptorNames: %v", err)
+	}
+	if len(values) != 0 {
+		t.Fatalf("There should be 0 Values, not %d", len(values))
+	}
+
+	v := models.ValueDescriptor{}
+	v.Id = id
+	v.Name = "name"
+	err = db.UpdateValueDescriptor(v)
+	if err != nil {
+		t.Fatalf("Error updating Value %v", err)
+	}
+	v2, err := db.ValueDescriptorById(v.Id.Hex())
+	if err != nil {
+		t.Fatalf("Error getting Value by id %v", err)
+	}
+	if v2.Name != v.Name {
+		t.Fatalf("Did not update Value correctly: %s %s", v.Name, v2.Name)
+	}
+
+	err = db.DeleteValueDescriptorById("INVALID")
+	if err == nil {
+		t.Fatalf("Value should not be deleted")
+	}
+
+	err = db.DeleteValueDescriptorById(id.Hex())
+	if err != nil {
+		t.Fatalf("Value should be deleted: %v", err)
+	}
+
+	err = db.UpdateValueDescriptor(v)
+	if err == nil {
+		t.Fatalf("Update should return error")
+	}
+}
+
+func testDB(t *testing.T, db DBClient) {
+	testDBReadings(t, db)
+	testDBEvents(t, db)
+	testDBValueDescriptors(t, db)
+
+	db.CloseSession()
+	db.CloseSession()
```

Comment:
This line should be removed

Reply:
I'm calling the ```CloseSession``` a second time to test that there is no panic when closing a db already closed. I will add a comment.

Reply:
ok! thanks for the clarification